docs/design/2020-04-17-sql-sequence.md

```diff
@@ -0,0 +1,118 @@
+
+- Usability: A sequence can be independent of tables, shared by multiple tables and be specified with CACHE details (either specify the number of CACHE or disable CACHE), which improves usability of the product.
+
+- Business Value: Sequence is a feature widely implemented in mainstream commercial databases. Users tend to expect TiDB can syntactically support sequence to reduce migration costs who migrate their business from other databases to TiDB. 
+
+## Background
+
+TiDB auto-increment will cache a batch values by default in each instance to ensure better performance, which means auto-increment only guarantee the uniqueness under a multi-tidb-node cluster. This kind of cache mode is uncontrollable and implicit in auto-increment.
+
+But in sequence, we take it open. Users can specify whether to use cache mode or not, what the cache size should be, what's the min and max of the range like and so on. This factors make sequence more flexible, controllable and easy to use.
+
+## Syntax and Semantics
+
+TiDB sequence is borrowed from MariaDB, which is also close to DB2 because DB2 is similar to MariaDB regarding the Sequence syntax.
+
+create sequence statement:
+
+```
+CREATE SEQUENCE [IF NOT EXISTS] sequence_name
+[ INCREMENT [ BY | = ] increment ]
+[ MINVALUE [=] minvalue | NO MINVALUE | NOMINVALUE ]
+[ MAXVALUE [=] maxvalue | NO MAXVALUE | NOMAXVALUE ]
+[ START [ WITH | = ] start ] 
+[ CACHE [=] cache | NOCACHE | NO CACHE ]
+[ CYCLE | NOCYCLE | NO CYCLE]
+```
+ 
+show create sequence statement:
+
+```
+SHOW CREATE SEQUENCE sequence_name
+```
+
+drop sequence sequence statement:
+
+```
+DROP SEQUENCE sequence_name_list
+```
+
+get the next value of sequence:
+
+```
+NEXT VALUE FOR sequence_name / NEXTVAL(sequence_name)
+```
+
+get the previous value of sequence:
+
+```
+LASTVAL(sequence_name)
+```
+
+move the current position in sequence number axis
+
+```
+SETVAL(sequence_name, num)
+```
+
+## Rationale
+
+- Sequence metadata
+
+Sequence is a table-level object, sharing the same namespace with the base table and view objects, whose meta are recorded in `ast.tableInfo`. But actually sequence definition do not have columns in it's meta, which means you can't operate the sequence value by `DML` interface. The only way you can do it is by `setval` function, after all, the sequence value is essentially a `KV` pair with some constraints.
+
+- Sequence values
+
+Each sequence value will be stored in its own range, with the key format below:
+
+```
+DB:<dbID> SID:<sequenceID> ——> int64 
+```
+
+Reads and writes to the sequence value (via the functions `nextval`, `setval`), will be implemented by direct calls to the `KV` layer's `Get`, `Inc`, and `Set` functions. Since sequence object store the value at the range of its own, it will cause hotpoint at sequence meta under insertion intensive cases.
```

Comment:
```suggestion
Reads and writes to the sequence value (via the functions `nextval`, `setval`), will be implemented by direct calls to the `KV` layer's `Get`, `Inc`, and `Set` functions. Since sequence objects store the value at the range of its own, it will cause hotspot at sequence meta under insertion intensive cases.
```